src/daemon/rpmostree-sysroot-upgrader.c

```diff
@@ -1040,6 +1050,186 @@ prep_local_assembly (RpmOstreeSysrootUpgrader *self,
+
+  g_auto(GLnxTmpfile) tmpf = { 0, };
+  if (!glnx_open_anonymous_tmpfile (O_RDWR | O_CLOEXEC, &tmpf, error))
+    return glnx_prefix_error (error, "Creating tmpfile");
+
+  g_autoptr(GOutputStream) tmpf_stream = g_unix_output_stream_new (tmpf.fd, FALSE);
+
+  g_autoptr(GPtrArray) filelist = g_ptr_array_new_with_free_func (g_free);
+
+  /* could do this async too... though it really shouldn't be that many files */
+  GLNX_HASH_TABLE_FOREACH (initramfs_etc_files, const char*, path)
+    {
+      /* should've been checked already */
+      g_assert (g_str_has_prefix (path, "/etc/"));
+      if (!add_etc_files_recurse (path, filelist, cancellable, error))
+        return FALSE;
+    }
+
+  /* feed into the input stream, but sort first to avoid needlessly changing the checksum */
+  g_ptr_array_sort (filelist, compare_strings);
+  g_autoptr(GInputStream) filelist_input = g_memory_input_stream_new ();
+
+  /* keep a hash of the previous path so we don't archive the same file multiple times */
+  guint last_path_hash = 0;
+  for (guint i = 0; i < filelist->len; i++)
+    {
+      g_autofree char *path = g_steal_pointer (&filelist->pdata[i]); /* steal in-place */
+      guint path_hash = g_str_hash (path);
+      if (path_hash == last_path_hash)
+        continue;
+      g_memory_input_stream_add_data (G_MEMORY_INPUT_STREAM (filelist_input),
+                                      g_steal_pointer (&path), (strlen (path))+1, g_free);
+      last_path_hash = path_hash;
+    }
+
+  const char *cpio_argv[] = {"cpio", "--create", "--format", "newc", "--quiet",
+                             "--reproducible", "--null", NULL};
+  g_autoptr(GSubprocess) cpio = g_subprocess_newv ((const char *const*)cpio_argv,
+                                                   G_SUBPROCESS_FLAGS_STDIN_PIPE |
+                                                   G_SUBPROCESS_FLAGS_STDOUT_PIPE,
+                                                   error);
+  if (!cpio)
+    return FALSE;
+
+  const char *gzip_argv[] = {"gzip", "-1", NULL};
+  g_autoptr(GSubprocess) gzip = g_subprocess_newv ((const char *const*)gzip_argv,
+                                                   G_SUBPROCESS_FLAGS_STDIN_PIPE |
+                                                   G_SUBPROCESS_FLAGS_STDOUT_PIPE,
+                                                   error);
+  if (!gzip)
+    return FALSE;
+
+  GOutputStream *cpio_input = g_subprocess_get_stdin_pipe (cpio);
+  g_output_stream_splice_async (cpio_input, filelist_input,
+                                G_OUTPUT_STREAM_SPLICE_CLOSE_SOURCE |
+                                G_OUTPUT_STREAM_SPLICE_CLOSE_TARGET,
+                                G_PRIORITY_DEFAULT, cancellable, splice_cb, &data);
+  data.counter++;
+
+  GInputStream *cpio_output = g_subprocess_get_stdout_pipe (cpio);
+  GOutputStream *gzip_input = g_subprocess_get_stdin_pipe (gzip);
+  g_output_stream_splice_async (gzip_input, cpio_output,
```

Comment:
It'd be more effficient and simpler to just pass cpio's stdout as gzip's stdin, right?